Mathlib/Computability/PrimrecFilter.lean

```diff
@@ -0,0 +1,179 @@
+- Primitive recursive functions are closed under bounded existential and universal quantifiers
+
+## References
+- [R. I. Soare *Turing Computability - Theory and Applications*] [Soare2016]
+-/
+
+open List Primrec
+
+namespace Primrec
+
+variable {α} [Primcodable α] (f : α → Prop) [DecidablePred f]
+
+/-- Filtering a list for elements that satisfy a decidable predicate is primitive recursive. -/
+lemma list_filter (hf : PrimrecPred f) : Primrec fun L ↦ (filter (fun a ↦ f a) L) := by
+  rw [← List.filterMap_eq_filter]
+  apply listFilterMap Primrec.id
+  simp only [Primrec₂, Option.guard, decide_eq_true_eq]
+  exact ite (PrimrecPred.comp hf snd) (option_some_iff.mpr snd) (const none)
+
+/-- Checking if any element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_exists (hf : PrimrecPred f) : PrimrecPred fun (L : List α) ↦ (∃ a ∈ L, f a) := by
+  let g := fun L ↦ List.filter (fun a ↦ f a) L
+  have h (L : List α): ((g L).length ≠ 0) ↔ (∃ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq (PrimrecPred.not ?_) h
+  exact PrimrecRel.comp Primrec.eq (comp list_length (list_filter f hf)) (const 0)
+
+/-- Checking if every element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_forall (hf : PrimrecPred f) : PrimrecPred fun (L : List α) ↦ (∀ a ∈ L, f a) := by
+  let g := fun L ↦ List.filter (fun a ↦ f a) L
+  have h (L : List α): ((g L).length = L.length) ↔ (∀ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq ?_ h
+  exact PrimrecRel.comp Primrec.eq (comp list_length (list_filter f hf)) list_length
+
+variable (f : ℕ → Prop) [DecidablePred f]
+
+/-- Bounded existential quantifiers are primitive recursive. -/
+lemma bounded_exists (hf : PrimrecPred f) : PrimrecPred fun n ↦ ∃ x < n, f x :=
+  PrimrecPred.of_eq (PrimrecPred.comp (filter_exists f hf) list_range) (by simp)
+
+/-- Bounded universal quantifiers are primitive recursive. -/
+lemma bounded_forall (hf : PrimrecPred f) : PrimrecPred fun n ↦ ∀ x < n, f x :=
+  PrimrecPred.of_eq (PrimrecPred.comp (filter_forall f hf) list_range) (by simp)
+
+end Primrec
+
+namespace primrec₂
+
+/-- If f a b is decidable, then given L : List α and b : β, it is primitive recurisve
+to filter L for elements a with f a b -/
+
+lemma filter_filtermap_lemma {α β : Type} (b : β) (f : α → β → Prop) (L : List α) [DecidableRel f] :
+    L.filter (fun a ↦ f a b) =
+    filterMap ((fun b a ↦ if f a b = True then some a else none) b) L := by
+  rw [← List.filterMap_eq_filter]
+  apply filterMap_congr
+  simp only [Option.guard, decide_eq_true_eq, eq_iff_iff, iff_true, implies_true]
+
+lemma list_filter {α β : Type} [Primcodable α] [Primcodable β] (f : α → β → Prop) [DecidableRel f]
+    (hf : PrimrecRel f) : Primrec₂ fun (L : List α) ↦ fun b ↦ (L.filter (fun a ↦ f a b)) := by
+  let g (b : β) : α → Option α := (fun a ↦ (if f a b = True then a else Option.none))
+  have h (b : β) (L : List α) := filter_filtermap_lemma b f L
+  simp only [h]
```

Comment:
You can simplify all these lines by `simp only [primrec₂.filter_filtermap_lemma]`.